File: src/main/java/ru/org/linux/edithistory/EditHistoryService.java
/*
 * Copyright 1998-2023 Linux.org.ru
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package ru.org.linux.edithistory;

import com.google.common.collect.ImmutableSet;
import com.google.common.collect.Iterables;
import org.springframework.stereotype.Service;
import ru.org.linux.comment.Comment;
import ru.org.linux.gallery.Image;
import ru.org.linux.gallery.ImageDao;
import ru.org.linux.gallery.ImageService;
import ru.org.linux.markup.MarkupType;
import ru.org.linux.markup.MessageTextService;
import ru.org.linux.poll.Poll;
import ru.org.linux.poll.PollDao;
import ru.org.linux.poll.PollNotFoundException;
import ru.org.linux.spring.dao.MessageText;
import ru.org.linux.spring.dao.MsgbaseDao;
import ru.org.linux.tag.TagName;
import ru.org.linux.tag.TagRef;
import ru.org.linux.tag.TagService;
import ru.org.linux.topic.PreparedImage;
import ru.org.linux.topic.Topic;
import ru.org.linux.topic.TopicTagService;
import ru.org.linux.user.User;
import ru.org.linux.user.UserDao;
import ru.org.linux.user.UserNotFoundException;
import ru.org.linux.user.UserService;
import scala.Option;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class EditHistoryService {
  private final TopicTagService topicTagService;

  private final UserDao userDao;

  private final UserService userService;

  private final MessageTextService textService;
  
  private final MsgbaseDao msgbaseDao;

  private final EditHistoryDao editHistoryDao;

  private final ImageDao imageDao;

  private final ImageService imageService;

  private final PollDao pollDao;

  public EditHistoryService(TopicTagService topicTagService, UserDao userDao, UserService userService,
                            MessageTextService textService, MsgbaseDao msgbaseDao, EditHistoryDao editHistoryDao,
                            ImageDao imageDao, ImageService imageService, PollDao pollDao) {
    this.topicTagService = topicTagService;
    this.userDao = userDao;
    this.userService = userService;
    this.textService = textService;
    this.msgbaseDao = msgbaseDao;
    this.editHistoryDao = editHistoryDao;
    this.imageDao = imageDao;
    this.imageService = imageService;
    this.pollDao = pollDao;
  }

  /**
   * Получить историю изменений топика
   */
  public List<PreparedEditHistory> prepareEditInfo(
    Topic topic,User currentUser
  ) throws UserNotFoundException {
    List<EditHistoryRecord> editInfoDTOs = editHistoryDao.getEditInfo(topic.getId(), EditHistoryObjectTypeEnum.TOPIC);
    List<PreparedEditHistory> editHistories = new ArrayList<>(editInfoDTOs.size());

    MessageText messageText = msgbaseDao.getMessageText(topic.getId());
    String currentMessage = messageText.text();
    MarkupType markup = messageText.markup();
    String currentTitle = topic.getTitle();
    String currentUrl = topic.getUrl();
    String currentLinktext = topic.getLinktext();
    List<TagRef> currentTags = topicTagService.getTagRefs(topic);
    boolean currentMinor = topic.isMinor();
    Image maybeImage = imageDao.imageForTopic(topic);
    PreparedImage currentImage = maybeImage !=null ? imageService.prepareImageJava(maybeImage).orElse(null) : null;
    Poll maybePoll;
    Integer lastId = null;

    try {
      maybePoll = pollDao.getPollByTopicId(topic.getId(),currentUser!=null ? currentUser.getId() : 0);
    } catch (PollNotFoundException ex) {
      maybePoll = null;
    }

    for (int i = 0; i < editInfoDTOs.size(); i++) {
      EditHistoryRecord dto = editInfoDTOs.get(i);

      editHistories.add(
        new PreparedEditHistory(
                textService,
          userDao.getUserCached(dto.getEditor()),
          dto.getEditdate(),
          dto.getOldmessage() != null ? currentMessage : null,
          dto.getOldtitle() != null ? currentTitle : null,
          dto.getOldurl() != null ? currentUrl : null,
          dto.getOldlinktext() != null ? currentLinktext : null,
          dto.getOldtags() != null ? currentTags : null,
          i == 0,
          false,
          dto.getOldminor() != null ? currentMinor : null,
          dto.getOldimage() != null && currentImage !=null ? currentImage : null,
          currentImage == null && dto.getOldimage()!=null,
          markup,
          dto.getOldPoll() !=null ? maybePoll : null,
          lastId
        )
      );

      if (dto.getOldimage() != null) {
        if (dto.getOldimage() == 0) {
          currentImage = null;
        } else {
          currentImage = imageService.prepareImageJava(imageDao.getImage(dto.getOldimage())).orElse(null);
        }
      }

      if (dto.getOldmessage() != null) {
        currentMessage = dto.getOldmessage();
        lastId = dto.getId();
      }

      if (dto.getOldtitle() != null) {
        currentTitle = dto.getOldtitle();
      }

      if (dto.getOldurl() != null) {
        currentUrl = dto.getOldurl();
      }

      if (dto.getOldlinktext() != null) {
        currentLinktext = dto.getOldlinktext();
      }

      if (dto.getOldtags() != null) {
        currentTags = TagService.namesToRefs(TagName.parseAndSanitizeTagsJava(dto.getOldtags()));
      }

      if (dto.getOldminor() != null) {
        currentMinor = dto.getOldminor();
      }

      if (dto.getOldPoll() != null) {
        maybePoll = dto.getOldPoll();
      }
    }

    if (!editInfoDTOs.isEmpty()) {
      if (currentTags.isEmpty()) {
        currentTags = null;
      }

      editHistories.add(new PreparedEditHistory(
              textService,
              userDao.getUserCached(topic.getAuthorUserId()),
              topic.getPostdate(),
              currentMessage,
              currentTitle,
              currentUrl,
              currentLinktext,
              currentTags,
              false,
              true,
              null,
              currentImage,
              false,
              markup,
              maybePoll,
              lastId));
    }

    return editHistories;
  }

  public List<PreparedEditHistory> prepareEditInfo(Comment comment) throws UserNotFoundException {
    List<EditHistoryRecord> editInfoDTOs = editHistoryDao.getEditInfo(comment.getId(), EditHistoryObjectTypeEnum.COMMENT);
    List<PreparedEditHistory> editHistories = new ArrayList<>(editInfoDTOs.size());

    MessageText messageText = msgbaseDao.getMessageText(comment.getId());
    MarkupType markup = messageText.markup();
    String currentMessage = messageText.text();
    String currentTitle = comment.getTitle();

    for (int i = 0; i < editInfoDTOs.size(); i++) {
      EditHistoryRecord dto = editInfoDTOs.get(i);

      editHistories.add(
        new PreparedEditHistory(
                textService,
          userDao.getUserCached(dto.getEditor()),
          dto.getEditdate(),
          dto.getOldmessage() != null ? currentMessage : null,
          dto.getOldtitle() != null ? currentTitle : null,
          null,
          null,
          null,
          i == 0,
          false,
          null,
                null, false, markup, null, null)
      );

      if (dto.getOldmessage() != null) {
        currentMessage = dto.getOldmessage();
      }

      if (dto.getOldtitle() != null) {
        currentTitle = dto.getOldtitle();
      }
    }

    if (!editInfoDTOs.isEmpty()) {
      editHistories.add(
        new PreparedEditHistory(
                textService,
          userDao.getUserCached(comment.getUserid()),
          comment.getPostdate(),
          currentMessage,
          currentTitle,
          null,
          null,
          null,
          false,
          true,
          null,
                null, false, markup, null, null)
      );
    }

    return editHistories;
  }

  public List<EditHistoryRecord> getEditInfo(int id, EditHistoryObjectTypeEnum objectTypeEnum) {
    return editHistoryDao.getEditInfo(id, objectTypeEnum);
  }

  public int editCount(int id, EditHistoryObjectTypeEnum objectTypeEnum) {
    // TODO replace with count() SQL query
    return editHistoryDao.getEditInfo(id, objectTypeEnum).size();
  }

  public void insert(EditHistoryRecord editHistoryRecord) {
    editHistoryDao.insert(editHistoryRecord);
  }

  public ImmutableSet<User> getEditorUsers(final Topic message, List<EditHistoryRecord> editInfoList) {
    ImmutableSet<Integer> editors = getEditors(message, editInfoList);

    return ImmutableSet.copyOf(userService.getUsersCachedJava(editors));
  }

  public ImmutableSet<Integer> getEditors(final Topic message, List<EditHistoryRecord> editInfoList) {
    return ImmutableSet.copyOf(
            Iterables.transform(
                    editInfoList.stream().filter(input -> input.getEditor() != message.getAuthorUserId()).collect(Collectors.toList()),
                    EditHistoryRecord::getEditor)
    );
  }

  public Option<EditInfoSummary> editInfoSummary(int id, EditHistoryObjectTypeEnum objectTypeEnum) {
    List<BriefEditInfo> history = editHistoryDao.getBriefEditInfo(id, objectTypeEnum);

    if (history.isEmpty()) {
      return Option.empty();
    } else {
      return Option.apply(EditInfoSummary.apply(history.size(), history.get(0)));
    }
  }

  public EditHistoryRecord getEditHistoryRecord(Topic topic, int recordId) {
    return editHistoryDao.getEditRecord(topic.getId(), recordId, EditHistoryObjectTypeEnum.TOPIC);
  }
}

File: src/main/scala/ru/org/linux/edithistory/EditHistoryService.scala
/*
 * Copyright 1998-2026 Linux.org.ru
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */
package ru.org.linux.edithistory

import cats.data.Chain
import org.springframework.stereotype.Service
import ru.org.linux.comment.Comment
import ru.org.linux.gallery.{ImageDao, ImageService}
import ru.org.linux.markup.{MarkupType, MessageTextService}
import ru.org.linux.msgbase.{MessageText, MsgbaseDao}
import ru.org.linux.poll.{Poll, PollDao, PollNotFoundException}
import ru.org.linux.scalikejdbc.Transaction
import ru.org.linux.tag.{TagRef, TagService}
import ru.org.linux.topic.{PreparedImage, Topic, TopicTagService}
import ru.org.linux.user.{User, UserService}

import java.sql.Timestamp
import java.util
import scala.jdk.CollectionConverters.SeqHasAsJava

@Service
class EditHistoryService(
    topicTagService: TopicTagService,
    userService: UserService,
    textService: MessageTextService,
    msgbaseDao: MsgbaseDao,
    editHistoryDao: EditHistoryDao,
    imageDao: ImageDao,
    imageService: ImageService,
    pollDao: PollDao):

  private case class TopicEditHistoryState(
      message: String,
      markup: MarkupType,
      title: String,
      url: String,
      linktext: String,
      tags: util.List[TagRef],
      minor: Boolean,
      lastId: Integer,
      poll: Poll,
      first: Boolean,
      images: Seq[PreparedImage]):
    def next(dto: EditHistoryRecord): TopicEditHistoryState =
      val images = dto
        .oldaddimages
        .map { images =>
          images.map(imageDao.getImage).flatMap(imageService.prepareImage)
        }
        .getOrElse(this.images)

      val (message, lastId) = dto
        .oldmessage
        .map { oldmessage =>
          (oldmessage, Integer.valueOf(dto.id))
        }
        .getOrElse((this.message, this.lastId))

      val title = dto.oldtitle.getOrElse(this.title)
      val url = dto.oldurl.getOrElse(this.url)
      val linktext = dto.oldlinktext.getOrElse(this.linktext)
      val tags = dto.oldtags.map(_.map(TagService.tagRef).asJava).getOrElse(this.tags)
      val minor = dto.oldminor.getOrElse(this.minor)
      val poll = dto.oldPoll.getOrElse(this.poll)

      this.copy(
        message = message,
        lastId = lastId,
        title = title,
        url = url,
        linktext = linktext,
        tags = tags,
        minor = minor,
        poll = poll,
        first = false,
        images = images
      )

    def build(dto: EditHistoryRecord): PreparedEditHistory =
      val (addedImages, removedImages, addedMainImage, removedMainImage) =
        if dto.oldaddimages.isDefined then
          val currentIds = this.images.map(_.image.id).toSet

          val added  = this.images.filterNot(img => dto.oldaddimages.get.contains(img.image.id))
          val removed = dto
            .oldaddimages
            .get
            .filterNot(currentIds.contains)
            .map(imageDao.getImage)
            .flatMap(imageService.prepareImage)

          (added.asJava, removed.asJava, null, null)

        else if dto.legacyMainImage.contains(0) then
          val addedMain =
            this.images.headOption.map(_ => this.images.take(1).toList.asJava).orNull

          (null, null, addedMain, null)

        else if dto.legacyMainImage.isDefined && dto.legacyMainImage.get > 0 then
          val xId = dto.legacyMainImage.get
          val removedMain =
            if this.images.exists(_.image.id == xId) then
              null
            else
              Seq(imageDao.getImage(xId)).flatMap(imageService.prepareImage).asJava

          (null, null, null, removedMain)

        else
          (null, null, null, null)

      val (message, restoreFrom) =
        if dto.oldmessage.isDefined then
          (textService.renderCommentText(MessageText(this.message, this.markup), false), this.lastId)
        else
          (null, null)

      PreparedEditHistory(
        original = false,
        editor = userService.getUserCached(dto.editor),
        message = message,
        current = this.first,
        title =
          if dto.oldtitle.isDefined then
            this.title
          else
            null
        ,
        tags =
          if dto.oldtags.isDefined then
            this.tags
          else
            null
        ,
        url =
          if dto.oldurl.isDefined then
            this.url
          else
            null
        ,
        linktext =
          if dto.oldlinktext.isDefined then
            this.linktext
          else
            null
        ,
        minor =
          if dto.oldminor.isDefined then
            this.minor
          else
            null
        ,
        editDate = new Timestamp(dto.editdate.toEpochMilli),
        poll =
          if dto.oldPoll.isDefined then
            this.poll
          else
            null
        ,
        restoreFrom = restoreFrom,
        addedImages = addedImages,
        removedImages = removedImages,
        addedMainImage = addedMainImage,
        removedMainImage = removedMainImage
      )

    def buildLast(topic: Topic): PreparedEditHistory =
      PreparedEditHistory(
        original = true,
        editor = userService.getUserCached(topic.authorUserId),
        message = textService.renderCommentText(MessageText(this.message, this.markup), false),
        current = false,
        title = this.title,
        tags =
          if !tags.isEmpty then
            this.tags
          else
            null
        ,
        url = this.url,
        linktext = this.linktext,
        minor = null,
        editDate = topic.postdate,
        poll = this.poll,
        restoreFrom = this.lastId,
        addedImages = this.images.asJava,
        removedImages = null,
        addedMainImage = null,
        removedMainImage = null
      )

  private object TopicEditHistoryState:
    def fromTopic(topic: Topic): TopicEditHistoryState =
      val messageText: MessageText = msgbaseDao.getMessageText(topic.id)
      val images = imageService.allImagesForTopic(topic)

      TopicEditHistoryState(
        message = messageText.text,
        markup = messageText.markup,
        title = topic.title,
        url = topic.url,
        linktext = topic.linktext,
        tags = topicTagService.getTagRefs(topic).asJava,
        minor = topic.minor,
        lastId = null,
        poll =
          try
            pollDao.getPollByTopicId(topic.id)
          catch
            case _: PollNotFoundException =>
              null
        ,
        first = true,
        images = images.flatMap(imageService.prepareImage)
      )

  private case class CommentEditHistoryState(message: String, markup: MarkupType, title: String, first: Boolean):
    def next(dto: EditHistoryRecord): CommentEditHistoryState =
      val message = dto.oldmessage.getOrElse(this.message)
      val title = dto.oldtitle.getOrElse(this.title)

      this.copy(first = false, message = message, title = title)

    def build(dto: EditHistoryRecord): PreparedEditHistory =
      val message =
        if dto.oldmessage.isDefined then
          textService.renderCommentText(MessageText(this.message, this.markup), false)
        else
          null

      PreparedEditHistory(
        original = false,
        editor = userService.getUserCached(dto.editor),
        message = message,
        current = this.first,
        title =
          if dto.oldtitle.isDefined then
            this.title
          else
            null
        ,
        tags = null,
        url = null,
        linktext = null,
        minor = null,
        editDate = new Timestamp(dto.editdate.toEpochMilli),
        poll = null,
        restoreFrom = null,
        addedImages = null,
        removedImages = null,
        addedMainImage = null,
        removedMainImage = null
      )

    def buildLast(comment: Comment): PreparedEditHistory =
      PreparedEditHistory(
        original = true,
        editor = userService.getUserCached(comment.userid),
        message = textService.renderCommentText(MessageText(this.message, this.markup), false),
        current = false,
        title = this.title,
        tags = null,
        url = null,
        linktext = null,
        minor = null,
        editDate = comment.postdate,
        poll = null,
        restoreFrom = null,
        addedImages = null,
        removedImages = null,
        addedMainImage = null,
        removedMainImage = null
      )

  private object CommentEditHistoryState:
    def fromComment(comment: Comment): CommentEditHistoryState =
      val messageText = msgbaseDao.getMessageText(comment.id)

      CommentEditHistoryState(
        markup = messageText.markup,
        message = messageText.text,
        title = comment.title,
        first = true)

  /** Получить историю изменений топика
    */
  def prepareEditInfo(topic: Topic): Seq[PreparedEditHistory] =
    val editInfoDTOs = editHistoryDao.getEditInfo(topic.id, EditHistoryObjectTypeEnum.TOPIC)

    if editInfoDTOs.nonEmpty then
      val initial = (TopicEditHistoryState.fromTopic(topic), Chain.empty[PreparedEditHistory])

      val (current, editHistories) =
        editInfoDTOs.foldLeft(initial) { case ((current, editHistories), dto) =>
          (current.next(dto), editHistories :+ current.build(dto))
        }

      (editHistories :+ current.buildLast(topic)).toVector
    else
      Vector.empty

  def prepareEditInfo(comment: Comment): Seq[PreparedEditHistory] =
    val editInfoDTOs = editHistoryDao.getEditInfo(comment.id, EditHistoryObjectTypeEnum.COMMENT)

    if editInfoDTOs.nonEmpty then
      val initial = (CommentEditHistoryState.fromComment(comment), Chain.empty[PreparedEditHistory])

      val (current, editHistories) =
        editInfoDTOs.foldLeft(initial) { case ((current, editHistories), dto) =>
          (current.next(dto), editHistories :+ current.build(dto))
        }

      (editHistories :+ current.buildLast(comment)).toVector
    else
      Vector.empty

  def getEditInfo(id: Int, objectTypeEnum: EditHistoryObjectTypeEnum): Seq[EditHistoryRecord] =
    editHistoryDao.getEditInfo(id, objectTypeEnum)

  def editCount(id: Int, objectTypeEnum: EditHistoryObjectTypeEnum): Int =
    // TODO replace with count() SQL query
    editHistoryDao.getEditInfo(id, objectTypeEnum).size

  def insert(editHistoryRecord: EditHistoryRecord)(using Transaction): Unit = editHistoryDao.insert(editHistoryRecord)

  def getEditorUsers(message: Topic, editInfoList: Seq[EditHistoryRecord]): Set[User] =
    val editors = editInfoList.view.map(_.editor).filterNot(_ == message.authorUserId).toSet

    userService.getUsersCached(editors).toSet

  def editInfoSummary(id: Int, objectTypeEnum: EditHistoryObjectTypeEnum): Option[EditInfoSummary] =
    val history = editHistoryDao.getBriefEditInfo(id, objectTypeEnum)

    history.headOption.map(v => EditInfoSummary(history.size, v))

  def getEditHistoryRecord(topic: Topic, recordId: Int): EditHistoryRecord =
    editHistoryDao.getEditRecord(topic.id, recordId, EditHistoryObjectTypeEnum.TOPIC)
